File: mobile/android/ActiveChainWallet/src/main/kotlin/dev/activechain/wallet/WalletBridge.kt
package dev.activechain.wallet

data class WalletIntentPreview(val recipient: String, val amount: Long, val feeReserve: Long,
                               val validUntil: Long, val policyAllowed: Boolean)

class LocalWalletBridge(private val dailyLimit: Long = 1_000) {
    fun previewTransfer(recipient: String, amount: Long, feeReserve: Long,
                        validUntil: Long, currentHeight: Long) =
        WalletIntentPreview(recipient, amount, feeReserve, validUntil,
            amount > 0 && amount <= dailyLimit && currentHeight <= validUntil)

    fun approveTransfer(preview: WalletIntentPreview): ByteArray {
        require(preview.policyAllowed) { "policy denied" }
        return "ACTIVECHAIN-LOCAL-INTENT-V1|${preview.recipient}|${preview.amount}|${preview.feeReserve}|${preview.validUntil}"
            .toByteArray(Charsets.UTF_8)
    }
}

File: mobile/ios/ActiveChainWallet/Sources/ActiveChainWallet/WalletBridge.swift
import Foundation

public struct WalletIntentPreview: Equatable {
    public let recipient: String
    public let amount: UInt64
    public let feeReserve: UInt64
    public let validUntil: UInt64
    public let policyAllowed: Bool
}

public protocol WalletBridge {
    func previewTransfer(recipient: String, amount: UInt64, feeReserve: UInt64,
                         validUntil: UInt64, currentHeight: UInt64) -> WalletIntentPreview
    func approveTransfer(_ preview: WalletIntentPreview) throws -> Data
}

public enum WalletError: Error { case policyDenied, expired }

/// Deterministic local shell used until the Rust FFI artifact is linked.
public final class LocalWalletBridge: WalletBridge {
    private let dailyLimit: UInt64
    public init(dailyLimit: UInt64 = 1_000) { self.dailyLimit = dailyLimit }

    public func previewTransfer(recipient: String, amount: UInt64, feeReserve: UInt64,
                                validUntil: UInt64, currentHeight: UInt64) -> WalletIntentPreview {
        WalletIntentPreview(recipient: recipient, amount: amount, feeReserve: feeReserve,
                            validUntil: validUntil,
                            policyAllowed: amount > 0 && amount <= dailyLimit && currentHeight <= validUntil)
    }

    public func approveTransfer(_ preview: WalletIntentPreview) throws -> Data {
        guard preview.policyAllowed else { throw WalletError.policyDenied }
        guard preview.amount > 0 else { throw WalletError.policyDenied }
        return Data("ACTIVECHAIN-LOCAL-INTENT-V1|\(preview.recipient)|\(preview.amount)|\(preview.feeReserve)|\(preview.validUntil)".utf8)
    }
}
